src/rust/engine/logging/src/logger.rs

```diff
@@ -185,12 +195,13 @@ impl Log for Logger {
         let log_string = format!("{} {} {}", time_str, level_marker, record.args());
 
         {
-          // If there are no handlers, or sending to any of the handlers failed, send to stderr
-          // directly.
+          // We first try to output to all registered handlers. If there are none, or any of them
+          // fail, then we fallback to sending directly to stderr.
           let handlers_map = self.stderr_handlers.lock();
           let mut any_handler_failed = false;
           for callback in handlers_map.values() {
-            if callback(&log_string).is_err() {
+            let handler_res = callback(&log_string);
+            if handler_res.is_err() {
```

Comment:
I made this change to try to make more obvious that `callback()` has a side effect.